File: src/main/java/data/stack/ValidateStackSequences_946.java
package data.stack;

import java.util.Stack;

public class ValidateStackSequences_946 {
  public static boolean validateStackSequences(int[] pushed, int[] popped) {
    var stack = new Stack<Integer>();
    var i = 0;
    for (var in : pushed) {
      stack.push(in);
      while (!stack.isEmpty() && stack.peek() == popped[i]) {
        stack.pop();
        i++;
      }
    }

    return stack.isEmpty();
  }

  public static void main(String[] args) {
    System.out.println(validateStackSequences(new int[]{1, 2, 3, 4, 5}, new int[]{4, 5, 3, 2, 1}));
    System.out.println(validateStackSequences(new int[]{1, 2, 3, 4, 5}, new int[]{4, 3, 5, 1, 2}));
  }
}

File: src/main/kotlin/data/structure/stack/ValidateStackSequences_946.kt
package data.structure.stack

import java.util.Stack

class KtValidateStackSequences_946 {
    fun validateStackSequences(pushed: IntArray, popped: IntArray): Boolean {
        val stack = Stack<Int>()
        var i = 0
        pushed.forEach { p ->
            stack.push(p)
            while (stack.isNotEmpty() && i < popped.size && stack.peek() == popped[i]) {
                stack.pop()
                i++
            }
        }
        return i == popped.size
    }
}

fun main() {
    val validate = KtValidateStackSequences_946()
    println(validate.validateStackSequences(intArrayOf(1,2,3,4,5), intArrayOf(4,5,3,2,1)))
}